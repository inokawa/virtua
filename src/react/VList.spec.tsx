import { afterEach, it, expect, describe, vitest } from "vitest";
import { render, cleanup } from "@testing-library/react";
import { VList } from "./VList";
import { Profiler, forwardRef } from "react";
import { CustomItemComponentProps } from "./types";
import { setupJsDomEnv } from "../../scripts/spec";

const ITEM_HEIGHT = 50;
const ITEM_WIDTH = 100;
const VIEWPORT_HEIGHT = ITEM_HEIGHT * 10;

setupJsDomEnv({
  itemWidth: ITEM_WIDTH,
  itemHeight: ITEM_HEIGHT,
  viewportHeight: VIEWPORT_HEIGHT,
});

afterEach(cleanup);

it("should pass attributes to element", () => {
  const { asFragment } = render(
    <VList
      id="id"
      className="class"
      tabIndex={0}
      role="list"
      aria-label="test"
      style={{ background: "red" }}
    >
      <div>0</div>
    </VList>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should pass index to items", () => {
  const Item = forwardRef<HTMLDivElement, CustomItemComponentProps>(
    ({ index, ...rest }, ref) => {
      return <div ref={ref} data-index={index} {...rest} />;
    }
  );
  const { asFragment } = render(
    <VList item={Item}>
      <div>0</div>
      <div>1</div>
      <div>2</div>
      <div>3</div>
      <div>4</div>
    </VList>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should render with render prop", () => {
  const items = Array.from({ length: 1000 }).map((_, i) => ({
    id: i,
    label: "This is " + i,
  }));
  const { asFragment } = render(
    <VList count={items.length}>
      {(i) => {
        const item = items[i]!;
        return <div key={item.id}>{item.label}</div>;
      }}
    </VList>
  );
  expect(asFragment()).toMatchSnapshot();
});

describe("vertical", () => {
  it("should render 0 children", () => {
    const { asFragment } = render(<VList>{[]}</VList>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 1 children", () => {
    const { asFragment } = render(
      <VList>
        <div>0</div>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 5 children", () => {
    const { asFragment } = render(
      <VList>
        <div>0</div>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 100 children", () => {
    const { asFragment } = render(
      <VList>
        {Array.from({ length: 100 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 1000 children", () => {
    const { asFragment } = render(
      <VList>
        {Array.from({ length: 1000 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 10000 children", () => {
    const { asFragment } = render(
      <VList>
        {Array.from({ length: 10000 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render non elements", () => {
    const { asFragment } = render(
      <VList>
        string
        {true}
        {false}
        {null}
        {undefined}
        {123}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render fragments", () => {
    const { asFragment } = render(
      <VList>
        <>
          <div>fragment</div>
          <div>fragment</div>
          <div>fragment</div>
        </>
        <>
          <div>fragment</div>
        </>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render component", () => {
    const Comp = ({ children }: { children: React.ReactNode }) => (
      <div>{children}</div>
    );
    const { asFragment } = render(
      <VList>
        <Comp>component</Comp>
        <Comp>component</Comp>
        <Comp>component</Comp>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with given width / height", () => {
    const { asFragment } = render(
      <VList style={{ width: 100, height: 800 }}>
        <div>0</div>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("horizontal", () => {
  it("should render 1 children", () => {
    const { asFragment } = render(
      <VList horizontal>
        <div>0</div>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 5 children", () => {
    const { asFragment } = render(
      <VList horizontal>
        <div>0</div>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 100 children", () => {
    const { asFragment } = render(
      <VList horizontal>
        {Array.from({ length: 100 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 1000 children", () => {
    const { asFragment } = render(
      <VList horizontal>
        {Array.from({ length: 1000 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 10000 children", () => {
    const { asFragment } = render(
      <VList horizontal>
        {Array.from({ length: 10000 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render non elements", () => {
    const { asFragment } = render(
      <VList horizontal>
        string
        {true}
        {false}
        {null}
        {undefined}
        {123}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render fragments", () => {
    const { asFragment } = render(
      <VList horizontal>
        <>
          <div>fragment</div>
          <div>fragment</div>
          <div>fragment</div>
        </>
        <>
          <div>fragment</div>
        </>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render component", () => {
    const Comp = ({ children }: { children: React.ReactNode }) => (
      <div>{children}</div>
    );
    const { asFragment } = render(
      <VList horizontal>
        <Comp>component</Comp>
        <Comp>component</Comp>
        <Comp>component</Comp>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with given width / height", () => {
    const { asFragment } = render(
      <VList horizontal style={{ width: 100, height: 800 }}>
        <div>0</div>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("reverse", () => {
  it("should render many items", () => {
    const { asFragment } = render(
      <VList reverse>
        {Array.from({ length: 100 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </VList>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("render count", () => {
  it("should render on mount", () => {
    const rootFn = vitest.fn();
    const itemCount = 4;
    const itemFns = Array.from({ length: itemCount }, (_) => vitest.fn());

    render(
      <Profiler id="root" onRender={rootFn}>
        <VList>
          {Array.from({ length: itemCount }, (_, i) => {
            const key = `item-${i}`;
            return (
              <Profiler key={key} id={key} onRender={itemFns[i]!}>
                <div>{i}</div>
              </Profiler>
            );
          })}
        </VList>
      </Profiler>
    );

    expect(rootFn).toBeCalledTimes(2);
    itemFns.forEach((itemFn) => {
      expect(itemFn).toHaveBeenCalledTimes(1);
    });
  });

  it("should render on mount many items", () => {
    const rootFn = vitest.fn();
    const itemCount = 100;
    const itemFns = Array.from({ length: itemCount }, (_) => vitest.fn());

    render(
      <Profiler id="root" onRender={rootFn}>
        <VList>
          {Array.from({ length: itemCount }, (_, i) => {
            const key = `item-${i}`;
            return (
              <Profiler key={key} id={key} onRender={itemFns[i]!}>
                <div>{i}</div>
              </Profiler>
            );
          })}
        </VList>
      </Profiler>
    );

    expect(rootFn).toBeCalledTimes(3);
    itemFns.forEach((itemFn) => {
      expect(itemFn.mock.calls.length).toBeLessThanOrEqual(1);
    });
  });

  // it("should render on length change", () => {
  //   let ready = false;
  //   const wrap = (f: (...args: any[]) => void) => {
  //     return (...args: any[]) => {
  //       if (!ready) return;
  //       f(...args);
  //     };
  //   };
  //   const rootFn = vitest.fn();

  //   const itemCount = 4;
  //   const itemFns = Array.from({ length: itemCount }, (_) => vitest.fn());

  //   const Mounter = ({
  //     children,
  //     count: initialCount,
  //   }: {
  //     children: (count: number) => ReactElement;
  //     count: number;
  //   }): ReactElement => {
  //     const [count, setCount] = useState(initialCount);
  //     useEffect(() => {
  //       ready = true;
  //       setCount((prev) => {
  //         return prev - 1;
  //       });
  //     }, []);
  //     return children(count);
  //   };

  //   render(
  //     <Mounter count={itemCount}>
  //       {(count) => (
  //         <Profiler id="root" onRender={wrap(rootFn)}>
  //           <VList>
  //             {Array.from({ length: count }, (_, i) => {
  //               const key = `item-${i}`;
  //               return (
  //                 <Profiler key={key} id={key} onRender={wrap(itemFns[i]!)}>
  //                   <div>{i}</div>
  //                 </Profiler>
  //               );
  //             })}
  //           </VList>
  //         </Profiler>
  //       )}
  //     </Mounter>
  //   );

  //   expect(rootFn).toBeCalledTimes(2);
  //   itemFns.forEach((itemFn, i) => {
  //     expect(itemFn).toBeCalledTimes(i === itemFns.length - 1 ? 0 : 1);
  //   });
  // });

  // it("should render on length change many items", () => {
  //   let ready = false;
  //   const wrap = (f: (...args: any[]) => void) => {
  //     return (...args: any[]) => {
  //       if (!ready) return;
  //       f(...args);
  //     };
  //   };
  //   const rootFn = vitest.fn();

  //   const itemCount = 100;
  //   const itemFns = Array.from({ length: itemCount }, (_) => vitest.fn());

  //   const Mounter = ({
  //     children,
  //     count: initialCount,
  //   }: {
  //     children: (count: number) => ReactElement;
  //     count: number;
  //   }): ReactElement => {
  //     const [count, setCount] = useState(initialCount);
  //     useEffect(() => {
  //       ready = true;
  //       setCount((prev) => {
  //         return prev - 1;
  //       });
  //     }, []);
  //     return children(count);
  //   };

  //   render(
  //     <Mounter count={itemCount}>
  //       {(count) => (
  //         <Profiler id="root" onRender={wrap(rootFn)}>
  //           <VList>
  //             {Array.from({ length: count }, (_, i) => {
  //               const key = `item-${i}`;
  //               return (
  //                 <Profiler key={key} id={key} onRender={wrap(itemFns[i]!)}>
  //                   <div>{i}</div>
  //                 </Profiler>
  //               );
  //             })}
  //           </VList>
  //         </Profiler>
  //       )}
  //     </Mounter>
  //   );

  //   expect(rootFn).toBeCalledTimes(3);
  //   itemFns.forEach((itemFn) => {
  //     expect(itemFn.mock.calls.length).toBeLessThanOrEqual(2); // TODO: should be 1
  //   });
  // });
});
