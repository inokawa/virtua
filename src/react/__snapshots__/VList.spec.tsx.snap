// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`horizontal > should render 1 children 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        <div>
          0
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`horizontal > should render 5 children 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 500px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 400px; visibility: visible; display: flex;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`horizontal > should render 100 children 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 10000px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 400px; visibility: visible; display: flex;"
      >
        <div>
          4
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 500px; visibility: visible; display: flex;"
      >
        <div>
          5
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`horizontal > should render 1000 children 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100000px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 400px; visibility: visible; display: flex;"
      >
        <div>
          4
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 500px; visibility: visible; display: flex;"
      >
        <div>
          5
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`horizontal > should render 10000 children 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 1000000px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 400px; visibility: visible; display: flex;"
      >
        <div>
          4
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 500px; visibility: visible; display: flex;"
      >
        <div>
          5
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`horizontal > should render component 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 300px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        <div>
          component
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: flex;"
      >
        <div>
          component
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: flex;"
      >
        <div>
          component
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`horizontal > should render fragments 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 200px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        <div>
          fragment
        </div>
        <div>
          fragment
        </div>
        <div>
          fragment
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: flex;"
      >
        <div>
          fragment
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`horizontal > should render non elements 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 200px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        string
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: flex;"
      >
        123
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`horizontal > should render with given width / height 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block; overflow-x: auto; contain: strict; width: 100px; height: 800px;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 500px; height: 100%;"
    >
      <div
        style="position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: flex;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: flex;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: flex;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: flex;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; height: 100%; top: 0px; left: 400px; visibility: visible; display: flex;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`reverse > should render many items 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="visibility: hidden; display: flex; flex-direction: column; justify-content: flex-end; min-height: 100%;"
    >
      <div
        style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 4000px;"
      >
        <div
          style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: hidden;"
        >
          <div>
            0
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 40px; visibility: hidden;"
        >
          <div>
            1
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 80px; visibility: hidden;"
        >
          <div>
            2
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 120px; visibility: hidden;"
        >
          <div>
            3
          </div>
        </div>
        <div
          style="position: absolute; width: 100%; left: 0px; top: 160px; visibility: hidden;"
        >
          <div>
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should pass attributes to element 1`] = `
<DocumentFragment>
  <div
    aria-label="test"
    class="class"
    id="id"
    role="list"
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%; background: red;"
    tabindex="0"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 50px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should pass index to items 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 250px;"
    >
      <div
        data-index="0"
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </div>
      <div
        data-index="1"
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          1
        </div>
      </div>
      <div
        data-index="2"
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          2
        </div>
      </div>
      <div
        data-index="3"
        style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
      >
        <div>
          3
        </div>
      </div>
      <div
        data-index="4"
        style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render with render prop 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 50000px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          This is 0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          This is 1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          This is 2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
      >
        <div>
          This is 3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
      >
        <div>
          This is 4
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 250px; visibility: visible;"
      >
        <div>
          This is 5
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render 0 children 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 0px;"
    />
  </div>
</DocumentFragment>
`;

exports[`vertical > should render 1 children 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 50px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render 5 children 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 250px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render 100 children 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 5000px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
      >
        <div>
          4
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 250px; visibility: visible;"
      >
        <div>
          5
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render 1000 children 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 50000px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
      >
        <div>
          4
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 250px; visibility: visible;"
      >
        <div>
          5
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render 10000 children 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 500000px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
      >
        <div>
          4
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 250px; visibility: visible;"
      >
        <div>
          5
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render component 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 150px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          component
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          component
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          component
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render fragments 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 100px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          fragment
        </div>
        <div>
          fragment
        </div>
        <div>
          fragment
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          fragment
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render non elements 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 100px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        string
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        123
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`vertical > should render with given width / height 1`] = `
<DocumentFragment>
  <div
    style="display: block; overflow-y: auto; contain: strict; width: 100px; height: 800px;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 250px;"
    >
      <div
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          1
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          2
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
      >
        <div>
          3
        </div>
      </div>
      <div
        style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
